File: updater.py
import json
import time
import subprocess
from datetime import datetime
import os


FILE = "data/commits.json"


def create_commit():

    now = datetime.now()

    message = (
        f"I committed at "
        f"{now.strftime('%Y-%m-%d %H:%M:%S')}"
    )


    new_commit = {

        "message": message,

        "time":
            now.strftime(
                "%Y-%m-%d %H:%M:%S"
            )

    }


    if os.path.exists(FILE):

        with open(FILE, "r") as file:
            commits = json.load(file)

    else:

        commits = []


    commits.append(new_commit)


    with open(FILE, "w") as file:

        json.dump(
            commits,
            file,
            indent=4
        )


    subprocess.run(
        ["git", "add", FILE]
    )


    subprocess.run(
        [
            "git",
            "commit",
            "-m",
            message
        ]
    )


    print(message)



if __name__ == "__main__":

    while True:

        create_commit()

        # Testing
        time.sleep(1)

        # For Jenkins demo:
        # time.sleep(60)

File: updater.js
const fs = require("fs");
const { execSync, spawn } = require("child_process");

const FILE = "data/commits.json";
const REMOTE = "https://github.com/Heiliger4/learn-jenkins-lab.git";

function run(command, options = {}) {
  return execSync(command, {
    encoding: "utf8",
    stdio: "inherit",
    ...options,
  });
}

function buildCommit(date = new Date()) {
  const timestamp = date
    .toISOString()
    .replace("T", " ")
    .substring(0, 19);

  return {
    message: `I committed at ${timestamp}`,
    time: timestamp,
  };
}

function appendCommit(commits, commit) {
  return [...commits, commit];
}

function createCommit() {
  const newCommit = buildCommit();

  // Create data directory if it doesn't exist
  if (!fs.existsSync("data")) {
    fs.mkdirSync("data", { recursive: true });
  }

  // Read existing commits
  let commits = [];

  if (fs.existsSync(FILE)) {
    try {
      commits = JSON.parse(fs.readFileSync(FILE, "utf8"));
    } catch {
      commits = [];
    }
  }

  const updatedCommits = appendCommit(commits, newCommit);

  // Save commits
  fs.writeFileSync(FILE, JSON.stringify(updatedCommits, null, 4));

  // Stage file
  run(`git add ${FILE}`);

  // Commit
  try {
    run(`git commit -m "${newCommit.message}"`);
  } catch {
    console.log("Nothing to commit.");
    return;
  }

  // Check remote
  try {
    let currentRemote = "";

    try {
      currentRemote = execSync(
        "git remote get-url origin",
        { encoding: "utf8" }
      ).trim();
    } catch {
      // Origin doesn't exist
    }

    if (!currentRemote) {
      run(`git remote add origin ${REMOTE}`);
      console.log(`Added remote origin -> ${REMOTE}`);
    } else if (currentRemote !== REMOTE) {
      run(`git remote set-url origin ${REMOTE}`);
      console.log(`Updated remote origin -> ${REMOTE}`);
    }

    // Get current branch
    const branch = execSync(
      "git rev-parse --abbrev-ref HEAD",
      { encoding: "utf8" }
    ).trim();

    // Push
    try {
      run(`git push -u origin ${branch}`);
      console.log(`Pushed to origin/${branch}`);
    } catch {
      console.log(
        "Push failed — check authentication and remote settings."
      );
    }
  } catch (error) {
    console.log("Error while pushing:", error.message);
  }

    console.log(newCommit.message);
}

module.exports = {
  buildCommit,
  appendCommit,
};

function runUpdaterLoop() {
  setInterval(() => {
    createCommit();
  }, 1000);
}

if (require.main === module) {
  runUpdaterLoop();
}
